File: solo/puzzles/medium/dice-probability-calculator/dice_probability_calculator.cpp
#include <iostream>
#include <string>
#include <map>
#include <vector>
#include <algorithm>
#include <stack>
#include <iomanip>

struct Distribution
{
   Distribution() = default;
   Distribution(int val)
   {
      outcomes[val] = 1.0;
   }

   Distribution applyOperator(const Distribution& other, char op) const
   {
      Distribution result;
      for (const auto& [a, pa] : outcomes)
      {
         for (const auto& [b, pb] : other.outcomes)
         {
            int r = 0;
            switch (op)
            {
            case '+':
               r = a + b;
               break;
            case '-':
               r = a - b;
               break;
            case '*':
               r = a * b;
               break;
            case '>':
               r = a > b;
               break;
            }
            result.outcomes[r] += pa * pb;
         }
      }
      return result;
   }

   std::map<int, double> outcomes;
};

Distribution die(int sides)
{
   Distribution d;
   for (int i = 1; i <= sides; ++i)
   {
      d.outcomes[i] = 1.0 / sides;
   }
   return d;
}

int precedence(const std::string& op)
{
   if (op == "*")
   {
      return 3;
   }
   else if (op == "+" || op == "-")
   {
      return 2;
   }
   else if (op == ">")
   {
      return 1;
   }
   return 0;
}

std::vector<std::string> reversePolishNotation(const std::string& expr)
{
   std::vector<std::string> reverse_polish_notation;
   std::stack<std::string> operators;

   int i = 0;
   while (i < expr.size())
   {
      if (isdigit(expr[i]) || expr[i] == 'd')
      {
         int j = i;
         while (j < expr.size() && (isdigit(expr[j]) || expr[j] == 'd'))
         {
            ++j;
         }
         reverse_polish_notation.push_back(expr.substr(i, j - i));
         i = j;
      }
      else if (expr[i] == '(')
      {
         operators.push(std::string(1, expr[i]));
         ++i;
      }
      else if (expr[i] == ')')
      {
         while (!operators.empty() && operators.top() != "(")
         {
            reverse_polish_notation.push_back(operators.top());
            operators.pop();
         }
         if (!operators.empty())
         {
            operators.pop();
         }
         ++i;
      }
      else
      {
         std::string op(1, expr[i]);
         while (!operators.empty() && precedence(operators.top()) >= precedence(op))
         {
            reverse_polish_notation.push_back(operators.top());
            operators.pop();
         }
         operators.push(op);
         ++i;
      }
   }
   while (!operators.empty())
   {
      reverse_polish_notation.push_back(operators.top());
      operators.pop();
   }
   return reverse_polish_notation;
}

Distribution evaluate(const std::vector<std::string>& rpn)
{
   std::stack<Distribution> stack;
   for (const std::string& token : rpn)
   {
      if (token.empty())
      {
         continue;
      }

      if (isdigit(token[0]))
      {
         stack.push(Distribution(std::stoi(token)));
      }
      else if (token[0] == 'd')
      {
         stack.push(die(std::stoi(token.substr(1))));
      }
      else
      {
         Distribution b = stack.top();
         stack.pop();
         Distribution a = stack.top();
         stack.pop();
         stack.push(a.applyOperator(b, token[0]));
      }
   }
   return stack.top();
}

int main()
{
   std::string expr;
   std::getline(std::cin, expr);

   std::map<int, int> results;

   std::vector<std::string> rpn = reversePolishNotation(expr);

   Distribution result = evaluate(rpn);

   std::vector<std::pair<int, double>> output(result.outcomes.begin(), result.outcomes.end());
   std::sort(output.begin(), output.end());

   for (const auto& [value, probability] : output)
   {
      std::cout << value << " " << std::fixed << std::setprecision(2) << probability * 100 << std::endl;
   }
}

File: solo/puzzles/medium/dice-probability-calculator/dice_probability_calculator.rs
use std::io;
use std::collections::HashMap;

macro_rules! parse_input {
   ($x:expr, $t:ident) => ($x.trim().parse::<$t>().unwrap())
}

struct Distribution {
   outcomes: HashMap<i32, f64>,
}

impl Distribution{
   pub fn new() -> Self {
      Distribution {
         outcomes: HashMap::new(),
      }
   }

   pub fn new_from_value(value: i32) -> Self {
      let mut outcomes = HashMap::new();
      outcomes.insert(value, 1.0);
      Distribution { outcomes }
   }

   pub fn new_from_dice(sides: i32) -> Self {
      let mut outcomes = HashMap::new();
      for i in 1..=sides {
         outcomes.insert(i, 1.0 / sides as f64);
      }
      Distribution { outcomes }
   }

   pub fn apply_operator(&self, other: &Distribution, operator: &str) -> Self {
      let mut new_outcomes = HashMap::new();
      for (a, a_probability) in &self.outcomes {
         for (b, b_probability) in &other.outcomes {
            let c = match operator {
               "+" => a + b,
               "-" => a - b,
               "*" => a * b,
               ">" => if a > b { 1 } else { 0 },
               _ => continue,
            };
            let new_probability = a_probability * b_probability;
            *new_outcomes.entry(c).or_insert(0.0) += new_probability;
         }
      }
      Distribution { outcomes: new_outcomes }
   }
}

struct Calculator {}

impl Calculator {
   pub fn new() -> Self {
      Calculator {}
   }

   fn precedence(op: &str) -> i32 {
      match op {
         ">" => 1,
         "+" | "-" => 2,
         "*" => 3,
         _ => 0,
      }
   }

   fn parse_to_rpn(&self, input: &str) -> Vec<String> {
      let mut output = Vec::new();
      let mut operators = Vec::new();
      let chars: Vec<char> = input.chars().collect();
      let mut i = 0;
      while i < chars.len() {
         if chars[i].is_ascii_digit() || chars[i] == 'd' {
            let mut j = i;
            while j < chars.len() && (chars[j].is_ascii_digit() || chars[j] == 'd') {
               j += 1;
            }
            output.push(input[i..j].to_string());
            i = j;
         } else if chars[i] == '(' {
            operators.push("(".to_string());
            i += 1;
         } else if chars[i] == ')' {
            while let Some(op) = operators.last() {
               if op == "(" {
                  break;
               }
               output.push(operators.pop().unwrap());
            }
            if operators.last() == Some(&"(".to_string()) {
               operators.pop();
            }
            i += 1;
         } else if "+-*>".contains(chars[i]) {
            let op = chars[i].to_string();
            while let Some(top) = operators.last() {
               if Self::precedence(top) >= Self::precedence(&op) {
                  output.push(operators.pop().unwrap());
               } else {
                  break;
               }
            }
            operators.push(op);
            i += 1;
         } else {
            i += 1;
         }
      }
      while let Some(op) = operators.pop() {
         output.push(op);
      }
      output
   }

   pub fn calculate(&self, expr: &str) -> Distribution {
      let rpn = self.parse_to_rpn(expr);
      let mut stack: Vec<Distribution> = Vec::new();
      for token in rpn {
         if token.chars().all(|c| c.is_ascii_digit()) {
            stack.push(Distribution::new_from_value(token.parse::<i32>().unwrap()));
         } else if token.starts_with('d') {
            let sides = token[1..].parse::<i32>().unwrap();
            stack.push(Distribution::new_from_dice(sides));
         } else {
            let right = stack.pop().unwrap();
            let left = stack.pop().unwrap();
            let result = left.apply_operator(&right, &token);
            stack.push(result);
         }
      }
      stack.pop().unwrap()
   }
}

fn main() {
   let mut input_line = String::new();
   io::stdin().read_line(&mut input_line).unwrap();
   let expr = input_line.trim_matches('\n').to_string();

   let calculator = Calculator{};
   let result = calculator.calculate(&expr);
   
   let mut outcomes: Vec<_> = result.outcomes.iter().collect();
   outcomes.sort_by_key(|&(outcome, _)| *outcome);
   for (outcome, probability) in outcomes {
      println!("{} {:.2}", outcome, probability * 100.0);
   }
}
